### PR TITLE
[Bugfix] Fix batch-invariant fp32 matmul OOR on SM89 for N=1

### PR DESCRIPTION
**Pull request by vhagor**
# [Bugfix] Fix batch-invariant fp32 matmul OOR on SM89 for N=1

---

## Purpose

On SM86/SM89 GPUs with a ~100 KB shared-memory budget (NVIDIA L20 / L4: **101376 B/block**), `VLLM_BATCH_INVARIANT=1` can fail during engine warmup:

```
triton.runtime.errors.OutOfResources: out of resource: shared memory,
Required: 131072, Hardware limit: 101376.
Reducing block sizes or `num_stages` may help.
```

SM8x installs Triton `matmul_persistent` for `aten::mm` / `addmm` / `matmul` / `linear`. Float32 used a single hardcoded tile (`128×128×32`, `num_stages=3`). fp16 already caps `BLOCK_SIZE_N` from shared memory (`#38670`); fp32 did not.

Triton still compiles a 128-wide N-tile and masks leftover columns. On L20 + Triton 3.6.0 it then specializes on the runtime `N`. For `N=1` that specialization adds ~32 KB of staging and the kernel needs **131072 B**. Sweep `N ∈ {1,2,…,4096}` × `K ∈ {768,2048}` (plus odd K, unaligned M, bmm): **only `N=1` OORs**.

Observed call chain on L20 warmup:

```
compile_or_warm_up_model
  -> dummy_pooler_run -> nn.Linear(..., 1)
  -> linear_batch_invariant -> matmul_batch_invariant
  -> matmul_persistent -> OutOfResources (131072 > 101376)
```

Fix: on CUDA devices with `get_max_shared_memory_bytes() <= 106496`, use `BLOCK_SIZE_N=32`, `num_stages=2` **only when `N == 1`**. Every other fp32 shape keeps `128 / stages=3`. `BLOCK_SIZE_K` stays 32. Tile choice depends on output width `N` (a layer property), not batch size `M`, so batch invariance is unchanged.

Not a duplicate of [#38670](https://github.com/vllm-project/vllm/pull/38670) (fp16 N-tile cap). Not a duplicate of open [#49131](https://github.com/vllm-project/vllm/pull/49131) (SM80 decode perf via **M**-dependent tiles; `M>64` keeps the original fp32 tile, which is exactly the warmup path that OORs).

## Change

In `enable_batch_invariant_mode()`:

| Device smem | fp16 `BLOCK_SIZE_N` | fp32 `N=1` tile |
|---|---|---|
| `> 106496` (e.g. A100) | 256 | `128 / stages=3` (unchanged) |
| `≤ 106496` (L20/L4) | 128 | `32 / stages=2` |

Launch (`matmul_persistent` and `bmm`): fp32 uses that N=1 override only when `N == 1`; `N > 1` stays `128 / stages=3`.

## Test Plan

```bash
.venv/bin/python -m pytest tests/v1/determinism/test_matmul_batch_invariant.py -v
.venv/bin/pre-commit run --files \
  vllm/model_executor/layers/batch_invariant.py \
  tests/v1/determinism/test_matmul_batch_invariant.py
```

New tests:

- `test_fp32_n_equals_1`: fp32 `N=1` matches `torch.mm` / `torch.bmm` and stays batch-invariant
- `test_fp32_n_gt_1_keeps_wide_tile`: `N>1` still matches `torch.mm` when the N=1 override is armed

L20 (SM 8.9, smem 101376), Triton 3.6.0:

1. Compile old vs new tile at `M∈{4,128}, K=2048, N=1`
2. N-sweep of the old tile
3. `VLLM_BATCH_INVARIANT=1` serve `BAAI/bge-reranker-base`; hit `/v1/models` and `/v1/rerank`

## Test Result

**Kernel compile (L20, Triton 3.6.0, old `128/stages=3`):**

| N | compiled shared | result |
|---:|---:|---|
| 1 | 131072 | OutOfResources (limit 101376) |
| 2–15 | 98304 | ok |
| ≥16 | 81920 | ok |

New N=1 tile (`32 / stages=2`): **36864 B**, launch ok (`M=4` and `M=128`).

## AI assistance

Drafting and code edits used Cursor. The submitting author reviewed every changed line and ran the L20 kernel / serve checks above.

### CODE REVIEW

File: tests/v1/determinism/test_matmul_batch_invariant.py [hunk 1]
```diff
@@ -103,3 +109,42 @@ def test_matmul_batch_invariance(dtype):
...
+def test_fp32_n_gt_1_keeps_wide_tile(monkeypatch: pytest.MonkeyPatch):
+    """N>1 must keep 128/stages=3 even when the N=1 override is armed."""
+    monkeypatch.setattr(batch_invariant_mod, "_fp32_block_size_n", 32)
+    monkeypatch.setattr(batch_invariant_mod, "_fp32_num_stages", 2)
+    device = torch.device(DEVICE_TYPE)
+    a = torch.randn(8, 2048, device=device, dtype=torch.float32)
+    b = torch.randn(2048, 16, device=device, dtype=torch.float32)
+    torch.testing.assert_close(
+        matmul_persistent(a, b), torch.mm(a, b), rtol=1e-2, atol=1e-2
+    )
```
Comment (yewentao256) [inline tests/v1/determinism/test_matmul_batch_invariant.py]:
```suggestion

```
I don't think we need a specific unit tests for this small update

File: vllm/model_executor/layers/batch_invariant.py [hunk 2]
```diff
@@ -174,10 +174,10 @@ def grid(META):
         },
         torch.float32: {
             "BLOCK_SIZE_M": 128,
-            "BLOCK_SIZE_N": 128,
+            "BLOCK_SIZE_N": _fp32_block_size_n if N == 1 else 128,
```
Comment (yewentao256) [inline vllm/model_executor/layers/batch_invariant.py:177]: What is the case that N will equal to 1, in production setting?

**Review by yewentao256 — COMMENTED** [review]
Thanks for the work!

File: vllm/model_executor/layers/batch_invariant.py (hunk 2 above)
Comment (vhagor) [reply vllm/model_executor/layers/batch_invariant.py:177]: Pooling mode + classify head (num_labels=1)

File: tests/v1/determinism/test_matmul_batch_invariant.py [hunk 3]
```diff
@@ -11,8 +11,14 @@
...
 
-from vllm.model_executor.layers.batch_invariant import matmul_batch_invariant
+from vllm.model_executor.layers import batch_invariant as batch_invariant_mod
+from vllm.model_executor.layers.batch_invariant import (
+    bmm_batch_invariant,
+    matmul_batch_invariant,
+    matmul_persistent,
+)
 from vllm.platforms import current_platform
+from vllm.utils.mem_utils import get_max_shared_memory_bytes
```
Comment (yewentao256) [inline tests/v1/determinism/test_matmul_batch_invariant.py]: Nit: idle import now

**Review by yewentao256 — APPROVED** [review]
LGTM, thanks for the work!
